src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -0,0 +1,132 @@
+    }
+    auto& aggregate_node = static_cast<AggregateNode&>(*node);
+
+    std::unordered_map<std::shared_ptr<const StoredTableNode>, std::set<ColumnID>> group_by_columns_per_table;
+    // Collect the group-by columns for each table in the aggregate node
+    for (auto expression_idx = size_t{0}; expression_idx < aggregate_node.aggregate_expressions_begin_idx;
+         ++expression_idx) {
+      const auto& expression = aggregate_node.node_expressions[expression_idx];
+      const auto& column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (!column_expression) continue;
+
+      const auto& stored_table_node =
+          std::dynamic_pointer_cast<const StoredTableNode>(column_expression->column_reference.original_node());
+      // If column is not a physical column skip
+      if (!stored_table_node) continue;
+
+      const auto column_id = column_expression->column_reference.original_column_id();
+
+      auto inserted = group_by_columns_per_table.try_emplace(stored_table_node, std::set<ColumnID>{column_id});
+      if (!inserted.second) {
+        group_by_columns_per_table[stored_table_node].insert(column_id);
+      }
+    }
+
+    // Store copy of aggregate expression to enable restoring the original order of columns via a projection
+    const auto initial_aggregate_column_expressions = aggregate_node.column_expressions();
+    bool group_by_list_changed = false;
+
+    // Main loop. Iterate over the tables and its group-by columns, gather primary keys and see if we can reduce.
+    for (const auto& [stored_table_node, group_by_columns] : group_by_columns_per_table) {
+      // Obtain column IDs of the primary key
+      auto unique_columns = std::set<ColumnID>();
+
+      const auto& table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+      if (table->get_soft_unique_constraints().empty()) {
+        // early exit for current table if no constraints are set
+        continue;
+      }
+
+      for (const auto& table_constraint : table->get_soft_unique_constraints()) {
+        if (table_constraint.is_primary_key == IsPrimaryKey::Yes) {
+          unique_columns.insert(table_constraint.columns.begin(), table_constraint.columns.end());
+          break;
+        }
+      }
+
+      // Intersect primary key and group-by columns. Only if full primary key is part of the group-by columns, the
+      // remaining columns can be removed.
+      std::vector<ColumnID> intersection;
+      std::set_intersection(unique_columns.begin(), unique_columns.end(), group_by_columns.begin(),
+                            group_by_columns.end(), std::back_inserter(intersection));
+
+      // Skip the current table as the primary key is not completely present.
+      if (intersection.size() != unique_columns.size()) {
+        continue;
+      }
+
+      for (const auto& group_by_column : group_by_columns) {
+        // Every column that is not part of the primary key is going to be removed.
```

Comment:
* that is part of the table but not part of the primary key
* removed? not adapted?